lib/IRGen/IRGenSIL.cpp

```diff
@@ -2802,38 +2804,124 @@ void IRGenSILFunction::visitGraphOperationInst(GraphOperationInst *i) {
+    // If we have an out parameter, then calculate the number of outputs
+    // required to fill it.
+    LLVM_DEBUG(llvm::dbgs() << " Counting outputs for indirect result of type "
+               << outParameterCanType << ".\n");
     auto *getTensorGroupCHandleCountFn =
         IGM.getTFC_GetTensorGroupCHandleCountFn();
     expectedReturnValueCount = Builder.CreateCall(
         getTensorGroupCHandleCountFn,
         {outParameterTypeMetadata, outParameterTypeMetadata,
          outParameterTensorGroupWitnessTable});
-    auto *allocateTensorGroupCHandleBufferFn =
-        IGM.getTFC_AllocateTensorGroupCHandleBufferFn();
-    returnValuesAddress = Builder.CreateCall(
-        allocateTensorGroupCHandleBufferFn,
-        {outParameterTypeMetadata, outParameterTypeMetadata,
-         outParameterTensorGroupWitnessTable});
-    returnValuesAddress = Builder.CreateBitCast(returnValuesAddress,
-                                                IGM.Int8PtrPtrTy);
+
+    // The calculation happens at runtime, so the output buffer must be heap
+    // allocated.
+    returnValuesHeapAllocated = true;
   } else {
-    // If we're returning results directly, we can tell how many results we have
-    // by looking at the instruction.
+    // If we're returning results directly, we must iterate over the results and
+    // add up their counts.
     LLVM_DEBUG(llvm::dbgs() << " Returns " << i->getNumResults()
                << " direct result(s).\n");
-    expectedReturnValueCount = llvm::ConstantInt::get(IGM.Int32Ty,
-                                                      i->getNumResults());
+    expectedReturnValueCount = llvm::ConstantInt::get(IGM.Int32Ty, 0);
+    for (auto silResult : i->getResults()) {
+      LLVM_DEBUG(llvm::dbgs() << "  Counting outputs for result of type "
+                 << silResult->getType() << ".\n");
+
+      // If the result is Void, it corresponds to 0 outputs.
+      if (silResult->getType().getASTType() == astCtx.TheEmptyTupleType) {
+        directResultTypeMetadatas.push_back(nullptr);
+        directResultTensorGroupWitnessTables.push_back(nullptr);
+        auto *cTensorHandleCount = llvm::ConstantInt::get(IGM.Int32Ty, 0);
+        directResultCTensorHandleCounts.push_back(cTensorHandleCount);
+        expectedReturnValueCount = Builder.CreateAdd(expectedReturnValueCount,
+                                                     cTensorHandleCount);
+        continue;
+      }
+
+      // If the result is a known TensorFlow type, it corresponds to just 1
+      // output.
+      // TODO: We could also handle concrete structs of known TensorFlow values
+      // here, to avoid falling through to the slower TensorGroup case.
+      if (tf::isTensorFlowValue(silResult->getType())) {
+        directResultTypeMetadatas.push_back(nullptr);
+        directResultTensorGroupWitnessTables.push_back(nullptr);
+        auto *cTensorHandleCount = llvm::ConstantInt::get(IGM.Int32Ty, 1);
```

Comment:
I tried out your code structure comment, and added a 2-pass algorithm. I think I like it, because it makes it more clear upfront whether we're going to be generating static code or doing a lot of runtime calls and heap allocations.

Regarding doing the counting at compile time, `llvm::Value` actually already does all the counting at compile time! There is some constant folding infrastructure in llvm's IRBuilder that keeps track of the `ConstantInt` value when you `CreateAdd` it with other `ConstantInt`s (even without any optimizations turned on). It only emits actual IR instructions when you add it with some runtime value.

Reply:
Thanks for the context. I suggest adding a comment here accordingly.

Also, would it be useful to add an LLVM IR-level compiler test (via FileChecker) to confirm the const folding has happened? Consider filing a (lower priority) bug to track this.

Reply:
I filed https://bugs.swift.org/browse/SR-9150